File: market_api.py
"""API helpers: SSL context, CLOB client builder, market lookup, resolution polling."""

from __future__ import annotations

import json
import os
import ssl
import sys
import time as _time
from datetime import datetime, timezone, timedelta

import requests

try:
    from py_clob_client.client import ClobClient
    from py_clob_client.clob_types import ApiCreds, BalanceAllowanceParams
    _HAS_PY_CLOB = True
except ImportError:
    _HAS_PY_CLOB = False

from market_config import MarketConfig

# ── SSL ──────────────────────────────────────────────────────────────────────
try:
    import certifi
    SSL_CTX = ssl.create_default_context(cafile=certifi.where())
except ImportError:
    SSL_CTX = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    SSL_CTX.check_hostname = False
    SSL_CTX.verify_mode = ssl.CERT_NONE

if sys.platform == "win32":
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")

# ── Endpoints ────────────────────────────────────────────────────────────────
GAMMA_API = "https://gamma-api.polymarket.com"
CLOB_HOST = "https://clob.polymarket.com"
CHAIN_ID = 137


# ── CLOB client builder ─────────────────────────────────────────────────────

def build_clob_client() -> ClobClient:
    """Build an authenticated ClobClient from .env credentials."""
    private_key = os.getenv("PRIVATE_KEY")
    if not private_key:
        print("  ERROR: PRIVATE_KEY not set in .env")
        sys.exit(1)

    funder = os.getenv("POLY_FUNDER", "")
    sig_type = int(os.getenv("SIGNATURE_TYPE", "1"))

    if not funder:
        print("  ERROR: POLY_FUNDER not set in .env")
        sys.exit(1)

    client = ClobClient(
        host=CLOB_HOST,
        key=private_key,
        chain_id=CHAIN_ID,
        signature_type=sig_type,
        funder=funder,
    )

    api_key = os.getenv("POLY_API_KEY", "")
    api_secret = os.getenv("POLY_API_SECRET", "")
    passphrase = os.getenv("POLY_PASSPHRASE", "")

    if api_key and api_secret and passphrase:
        client.set_api_creds(ApiCreds(
            api_key=api_key,
            api_secret=api_secret,
            api_passphrase=passphrase,
        ))
    else:
        print("  Deriving API credentials...")
        creds = client.create_or_derive_api_creds()
        client.set_api_creds(creds)
        print(f"  Save these in .env:")
        print(f"    POLY_API_KEY={creds.api_key}")
        print(f"    POLY_API_SECRET={creds.api_secret}")
        print(f"    POLY_PASSPHRASE={creds.api_passphrase}")

    return client


def query_usdc_balance(client: ClobClient, debug: bool = False) -> float | None:
    """Fetch current USDC (collateral) balance from the CLOB API."""
    try:
        resp = client.get_balance_allowance(
            BalanceAllowanceParams(asset_type="COLLATERAL")
        )
        raw = float(resp.get("balance", 0))
        if raw > 1_000_000:
            return raw / 1e6
        return raw
    except Exception as exc:
        if debug:
            print(f"  [BALANCE] error: {exc}")
        return None


# ── Helpers ──────────────────────────────────────────────────────────────────

def _ensure_list(val):
    return json.loads(val) if isinstance(val, str) else val


def _try_slug(slug: str):
    try:
        resp = requests.get(
            f"{GAMMA_API}/events", params={"slug": slug}, timeout=10
        )
        data = resp.json()
        if data:
            return data[0], data[0]["markets"][0]
    except Exception:
        pass
    return None


def _generate_hourly_slug(config: MarketConfig, target_utc: datetime) -> str:
    """Generate a human-readable slug for 1-hour markets.

    Hourly markets use ET-based slugs like:
      bitcoin-up-or-down-april-9-2026-4pm-et

    The slug_prefix for 1h configs is the asset name part
    (e.g. "bitcoin-up-or-down"), and we append the date + hour in ET.
    """
    from zoneinfo import ZoneInfo
    et = target_utc.astimezone(ZoneInfo("America/New_York"))
    month = et.strftime("%B").lower()  # "april"
    day = et.day
    year = et.year
    hour = et.hour
    if hour == 0:
        time_str = "12am"
    elif hour < 12:
        time_str = f"{hour}am"
    elif hour == 12:
        time_str = "12pm"
    else:
        time_str = f"{hour - 12}pm"
    return f"{config.slug_prefix}-{month}-{day}-{year}-{time_str}-et"


def find_market(config: MarketConfig):
    now = datetime.now(timezone.utc)
    align = config.window_align_m

    # Hourly markets use human-readable ET-based slugs instead of
    # timestamp-based slugs. Detect by window duration.
    is_hourly = config.window_duration_s >= 3600

    if is_hourly:
        # Try current hour, then next, then previous (in case we're
        # near a boundary and the current window just closed).
        hour_start = now.replace(minute=0, second=0, microsecond=0)
        for offset_h in [0, 1, -1, 2]:
            candidate = hour_start + timedelta(hours=offset_h)
            slug = _generate_hourly_slug(config, candidate)
            result = _try_slug(slug)
            if result:
                event, market = result
                end = datetime.fromisoformat(
                    market["endDate"].replace("Z", "+00:00")
                )
                if now < end:
                    return event, market
    else:
        # Original timestamp-based slug search for 5m/15m markets
        minute = (now.minute // align) * align
        window_start = now.replace(minute=minute, second=0, microsecond=0)

        for offset in [0, -align, align, -2 * align, 2 * align]:
            candidate = window_start + timedelta(minutes=offset)
            ts = int(candidate.timestamp())
            result = _try_slug(f"{config.slug_prefix}-{ts}")
            if result:
                event, market = result
                end = datetime.fromisoformat(
                    market["endDate"].replace("Z", "+00:00")
                )
                start = datetime.fromisoformat(
                    market["eventStartTime"].replace("Z", "+00:00")
                )
                if now < end or start > now:
                    return event, market

    # Broad fallback: search active up-or-down markets
    try:
        resp = requests.get(
            f"{GAMMA_API}/events",
            params={
                "active": "true", "closed": "false",
                "tag_slug": "up-or-down", "limit": 100,
            },
            timeout=15,
        )
        for e in resp.json():
            if config.slug_prefix not in e.get("slug", ""):
                continue
            m = e["markets"][0]
            end = datetime.fromisoformat(m["endDate"].replace("Z", "+00:00"))
            if now < end:
                return e, m
    except Exception:
        pass

    return None, None


def poll_market_resolution(slug: str, max_attempts: int = 12,
                           delay: float = 5.0,
                           debug: bool = False) -> int | None:
    """Poll Gamma API until the market is FULLY RESOLVED (not just closed).

    2026-04-09 fix: previously this checked only `closed=True` and read
    `outcomePrices` as a winner indicator (`up_price > 0.5`). But the API
    sets `closed=True` BEFORE the on-chain resolution finalizes, and during
    that gap `outcomePrices` shows the LAST TRADE PRICES (not the
    settlement prices). A market that traded UP at 0.51 just before
    closing would be reported as "UP wins" by the old code, even when the
    actual Chainlink resolution was DOWN. Real loss documented:
    btc-updown-5m-1775709300 (window 04:35-04:40 UTC, 2026-04-09) — bot
    bet DOWN, recorded as UP loss, actual API state now shows DOWN won
    (outcomePrices=["0","1"], umaResolutionStatus="resolved"). Cost: $9.30
    in unredeemed winnings.

    The fix requires THREE conditions before trusting the outcome:
    1. `closed == True`
    2. `umaResolutionStatus == "resolved"` (UMA optimistic oracle confirmed)
    3. `outcomePrices` is exactly one "1" and one "0" (settlement prices)

    If any condition fails, keep polling (up to max_attempts).
    """
    for attempt in range(max_attempts):
        try:
            resp = requests.get(
                f"{GAMMA_API}/events", params={"slug": slug}, timeout=10
            )
            data = resp.json()
            if not data:
                _time.sleep(delay)
                continue
            market = data[0]["markets"][0]
            if not market.get("closed"):
                if debug:
                    print(f"  [RESOLVE] attempt {attempt + 1}: not closed yet")
                _time.sleep(delay)
                continue

            # NEW: require UMA resolution before trusting outcomePrices.
            uma_status = market.get("umaResolutionStatus", "")
            if uma_status != "resolved":
                if debug:
                    print(f"  [RESOLVE] attempt {attempt + 1}: closed but uma_status={uma_status!r} (waiting for finalization)")
                _time.sleep(delay)
                continue

            outcomes = _ensure_list(market["outcomes"])
            outcome_prices = _ensure_list(market["outcomePrices"])
            if len(outcomes) != 2 or len(outcome_prices) != 2:
                if debug:
                    print(f"  [RESOLVE] attempt {attempt + 1}: malformed outcomes/prices")
                _time.sleep(delay)
                continue

            # NEW: require settlement prices to be exactly 0/1.
            # If the market is showing intermediate prices like 0.51/0.49,
            # the on-chain resolution hasn't been applied yet — keep polling.
            try:
                price_strs = [str(p).strip() for p in outcome_prices]
                price_floats = [float(p) for p in outcome_prices]
            except (ValueError, TypeError):
                if debug:
                    print(f"  [RESOLVE] attempt {attempt + 1}: price parse fail: {outcome_prices}")
                _time.sleep(delay)
                continue

            settled = (
                {price_strs[0], price_strs[1]} == {"0", "1"}
                or sorted(price_floats) == [0.0, 1.0]
            )
            if not settled:
                if debug:
                    print(f"  [RESOLVE] attempt {attempt + 1}: prices not settled yet: {outcome_prices}")
                _time.sleep(delay)
                continue

            up_idx = outcomes.index("Up")
            up_price = float(outcome_prices[up_idx])
            return 1 if up_price >= 1.0 else 0
        except Exception as exc:
            if debug:
                print(f"  [RESOLVE] attempt {attempt + 1} error: {exc}")
            _time.sleep(delay)
    return None

File: rust/src/market_api.rs
use anyhow::{anyhow, Result};
use chrono::{DateTime, Duration, Timelike, Utc};
use serde_json::Value;

use crate::config::MarketConfig;

pub const GAMMA_API: &str = "https://gamma-api.polymarket.com";
pub const CLOB_HOST: &str = "https://clob.polymarket.com";
pub const CHAIN_ID: u64 = 137;

#[derive(Clone, Debug)]
pub struct Market {
    pub slug: String,
    pub up_token: String,
    pub down_token: String,
    pub end_time: DateTime<Utc>,
    pub start_time: DateTime<Utc>,
    pub condition_id: String,
}

fn ensure_list(v: &Value) -> Value {
    match v {
        Value::String(s) => serde_json::from_str(s).unwrap_or(Value::Null),
        other => other.clone(),
    }
}

fn parse_iso(s: &str) -> Result<DateTime<Utc>> {
    let s = s.replace('Z', "+00:00");
    Ok(DateTime::parse_from_rfc3339(&s)?.with_timezone(&Utc))
}

fn parse_market(event: &Value) -> Result<Market> {
    let markets = event
        .get("markets")
        .and_then(|v| v.as_array())
        .ok_or_else(|| anyhow!("no markets array"))?;
    let m = markets.first().ok_or_else(|| anyhow!("empty markets"))?;

    let slug = event.get("slug").and_then(|v| v.as_str()).unwrap_or("").to_string();
    let end_time = parse_iso(m.get("endDate").and_then(|v| v.as_str()).unwrap_or_default())?;
    let start_time = parse_iso(
        m.get("eventStartTime")
            .and_then(|v| v.as_str())
            .or_else(|| m.get("startDate").and_then(|v| v.as_str()))
            .unwrap_or_default(),
    )
    .unwrap_or(end_time - Duration::seconds(900));
    let condition_id = m
        .get("conditionId")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let tokens = m.get("clobTokenIds").map(ensure_list).unwrap_or(Value::Null);
    let ids: Vec<String> = tokens
        .as_array()
        .map(|a| a.iter().filter_map(|v| v.as_str().map(String::from)).collect())
        .unwrap_or_default();
    if ids.len() < 2 {
        return Err(anyhow!("need 2 clobTokenIds, got {}", ids.len()));
    }

    let outcomes = m.get("outcomes").map(ensure_list).unwrap_or(Value::Null);
    let outcome_strs: Vec<String> = outcomes
        .as_array()
        .map(|a| a.iter().filter_map(|v| v.as_str().map(String::from)).collect())
        .unwrap_or_default();
    let up_idx = outcome_strs
        .iter()
        .position(|o| o.eq_ignore_ascii_case("Up"))
        .unwrap_or(0);
    let down_idx = 1 - up_idx;

    Ok(Market {
        slug,
        up_token: ids[up_idx].clone(),
        down_token: ids[down_idx].clone(),
        end_time,
        start_time,
        condition_id,
    })
}

async fn try_slug(client: &reqwest::Client, slug: &str) -> Option<Value> {
    let url = format!("{GAMMA_API}/events");
    let resp = client.get(&url).query(&[("slug", slug)]).send().await.ok()?;
    let data: Value = resp.json().await.ok()?;
    let arr = data.as_array()?;
    arr.first().cloned()
}

pub async fn find_market(client: &reqwest::Client, config: &MarketConfig) -> Result<Market> {
    let now = Utc::now();
    let align = config.window_align_m;
    let minute = (now.minute() as i64 / align) * align;
    let window_start = now
        .with_minute(minute as u32)
        .unwrap()
        .with_second(0)
        .unwrap()
        .with_nanosecond(0)
        .unwrap();

    for offset in [0, -align, align, -2 * align, 2 * align] {
        let candidate = window_start + Duration::minutes(offset);
        let ts = candidate.timestamp();
        let slug = format!("{}-{}", config.slug_prefix, ts);
        if let Some(event) = try_slug(client, &slug).await {
            if let Ok(market) = parse_market(&event) {
                if now < market.end_time || market.start_time > now {
                    return Ok(market);
                }
            }
        }
    }

    // Fallback: search active up-or-down markets
    let url = format!("{GAMMA_API}/events");
    let resp = client
        .get(&url)
        .query(&[
            ("active", "true"),
            ("closed", "false"),
            ("tag_slug", "up-or-down"),
            ("limit", "100"),
        ])
        .send()
        .await?;
    let data: Value = resp.json().await?;
    if let Some(arr) = data.as_array() {
        for event in arr {
            let slug = event.get("slug").and_then(|v| v.as_str()).unwrap_or("");
            if !slug.contains(config.slug_prefix) {
                continue;
            }
            if let Ok(market) = parse_market(event) {
                if now < market.end_time {
                    return Ok(market);
                }
            }
        }
    }

    Err(anyhow!("no active market for {}", config.slug_prefix))
}

/// Poll Gamma until the market is fully resolved, returning 1 (UP won) or 0 (DOWN won).
///
/// Requires all three:
///   - closed = true
///   - umaResolutionStatus = "resolved"
///   - outcomePrices is exactly {"0","1"} (or {0.0, 1.0})
pub async fn poll_resolution(
    client: &reqwest::Client,
    slug: &str,
    max_attempts: u32,
    delay_s: f64,
) -> Option<u8> {
    for _ in 0..max_attempts {
        if let Ok(resp) = client
            .get(format!("{GAMMA_API}/events"))
            .query(&[("slug", slug)])
            .send()
            .await
        {
            if let Ok(data) = resp.json::<Value>().await {
                if let Some(event) = data.as_array().and_then(|a| a.first()) {
                    if let Some(market) = event
                        .get("markets")
                        .and_then(|v| v.as_array())
                        .and_then(|a| a.first())
                    {
                        let closed = market.get("closed").and_then(|v| v.as_bool()).unwrap_or(false);
                        let uma = market
                            .get("umaResolutionStatus")
                            .and_then(|v| v.as_str())
                            .unwrap_or("");
                        if closed && uma == "resolved" {
                            let prices = ensure_list(
                                market.get("outcomePrices").unwrap_or(&Value::Null),
                            );
                            let outcomes = ensure_list(
                                market.get("outcomes").unwrap_or(&Value::Null),
                            );
                            if let (Some(p_arr), Some(o_arr)) =
                                (prices.as_array(), outcomes.as_array())
                            {
                                let pfloats: Vec<f64> = p_arr
                                    .iter()
                                    .filter_map(|v| {
                                        v.as_str().and_then(|s| s.parse().ok()).or_else(|| v.as_f64())
                                    })
                                    .collect();
                                if pfloats.len() == 2 {
                                    let mut sorted = pfloats.clone();
                                    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
                                    if sorted == [0.0, 1.0] {
                                        let up_idx = o_arr
                                            .iter()
                                            .position(|o| {
                                                o.as_str()
                                                    .map(|s| s.eq_ignore_ascii_case("Up"))
                                                    .unwrap_or(false)
                                            })
                                            .unwrap_or(0);
                                        return Some(if pfloats[up_idx] >= 0.5 { 1 } else { 0 });
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
        tokio::time::sleep(std::time::Duration::from_secs_f64(delay_s)).await;
    }
    None
}

